Resampling (chunk 1/3) (#44)
aml4td/website: chapters/resampling.qmd

library(tidymodels)
library(patchwork)
library(kableExtra)
library(doParallel)
library(leaflet)
library(spatialsample)

# Also requires: 
# for model fitting
# gamair, viridisLite

# ------------------------------------------------------------------------------
# Set options
cl <- makePSOCKcluster(parallel::detectCores(logical = TRUE))
registerDoParallel(cl)
tidymodels_prefer()
theme_set(theme_transparent())
set_options()

# ------------------------------------------------------------------------------
# Load previous data sets

load("../RData/resampling_var_bias.RData")
sel_model <- "logistic"
sel_metric <- "brier_class"

metric_data <-
  bind_rows(
    stats_bootstraps %>% filter(.metric == sel_metric),
    stats_mc_cv %>% filter(.metric == sel_metric),
    stats_v_fold_cv %>% filter(.metric == sel_metric)
  ) %>%
  filter(model == sel_model)

rng_bias <- range(c(metric_data$bias, 0))
rng_std_err <- range(metric_data$std_err, na.rm = TRUE)
rng_std_err[1] <- 0

br_lr <- 
  stats_bootstraps %>% 
  filter(model == "logistic" & .metric == "brier_class" & times == 100)

br_1nn <- 
  stats_bootstraps %>% 
  filter(model == "knn" & .metric == "brier_class" & times == 100)
```

> In @sec-external-validation, it was shown that using the same data to estimate and evaluate our model can produce inaccurate estimates of how well it functions. It also described using separate partitions of the data to fit and evaluate the model. 

Reviewer: ```suggestion
In @sec-external-validation, it was shown that using the same data to estimate and evaluate our model can produce inaccurate estimates of how well it functions to predict new samples. It also described using separate partitions of the data to fit and evaluate the model. 
```

> There are two general approaches for using external data for model evaluation. The first is a validation set, which we’ve seen before in @sec-three-way-split This is a good idea if you have a lot of data on hand. The second approach is to **resample** the training set. Resampling is an iterative approach that reuses the training data multiple times using sound statistical methodology. We’ll discuss both validation sets and resampling in this chapter. 

Reviewer: ```suggestion
There are two general approaches for using external data for model evaluation. The first is a validation set, which we’ve seen before in @sec-three-way-split. This is a good idea if you have a lot of data on hand. The second approach is to **resample** the training set. Resampling is an iterative approach that reuses the training data multiple times based on sound statistical methodology. We’ll discuss both validation sets and resampling in this chapter. 
```

There have been ongoing efforts to understand what resampling methods do on a theoretical level. @bates2023cross is an interesting work focused on linear regression models with ordinary least squares. Their conclusion regarding resampling estimates is that they

> ... cannot be viewed as estimates of the prediction error of the final model fit on the whole data. Rather, the estimate of prediction error is an estimate of the average prediction error of the final model across other hypothetical data sets from the same distribution.

It is a good idea to think that this will also be true for more complex machine learning models. The important idea is that resampling is measuring performance on training sets similar to ours (and the same size). 

To understand different methods, let’s examine some theoretical properties of resampling estimates that matter to applied data analysis. 

> Let’s start with analogy; consider the usual sample mean estimator used in basic statistics ($\bar{x}$). Based on assumptions about the data, we can derive an equation that optimally estimates the true mean value based on some criterion. Often, the statistical theory focuses on our estimators' theoretical mean and variance. For example, if the data are independent and follow the same Gaussian distribution, the sample mean attempts to estimate the actual population mean (i.e., it is an **unbiased** estimator). We can sometimes derive what the estimator's theoretical variance is, too. These properties, or some combination of them, such as mean squared error^[This discussion is very similar to the variance-bias tradeoff discussed in @sec-variance-bias. The context here differs, but the themes are the same.], help guide us to the best estimator. 

Reviewer: ```suggestion
Let’s start with analogy; consider the usual sample mean estimator used in basic statistics ($\bar{x}$). Based on assumptions about the data, we can derive an equation that optimally estimates the true mean value based on some criterion. Often, the statistical theory focuses on our estimators' theoretical mean and variance. For example, if the data are independent and follow the same Gaussian distribution, the sample mean attempts to estimate the actual population mean (i.e., it is an **unbiased** estimator). Under the same assumptions we can also derive what the estimator's theoretical variance. These properties, or some combination of them, such as mean squared error^[This discussion is very similar to the variance-bias tradeoff discussed in @sec-variance-bias. The context here differs, but the themes are the same.], help guide us to the best estimator. 
```

> The same is true for resampling methods. We can understand how their bias and variances change under different circumstances and choose an appropriate technique. We'll also focus on bias and variance as the main properties of interest. 

Reviewer: ```suggestion
The same is true for resampling methods. We can understand how their bias and variance change under different circumstances and choose an appropriate technique.  To demonstrate resampling methods, we'll focus on bias and variance as the main properties of interest. 
```

> First is bias: how accurately does a resampling technique estimate the true population value $Q$? We'd like it to be unbiased, but to our knowledge, that is nearly impossible. However, some things we do know.  @fig-resampling-scheme shows that the training set (of size $n_{tr}$) is split into two partitions of size $n_{fit}$ and $n_{pred}$. It turns out that, as $n_{pred}$ becomes larger, the bias increases in a pessimistic direction^[Meaning that performance looks worse than it should. For example, smaller $R^2$ or inflated Brier score.]. In other words, if our training set has 1,000 samples, a resampling method where the assessment set has $n_{pred} = 100$ data points has a smaller bias than one using $n_{pred} = 500$. Another thing that we know is that increasing the number of resamples (_B_) can't significantly reduce the bias. 

Reviewer: ```suggestion
First is bias: how accurately does a resampling technique estimate the true population value $Q$? We'd like it to be unbiased, but to our knowledge, that is nearly impossible. However, some things we do know.  @fig-resampling-scheme shows that the training set (of size $n_{tr}$) is split into _B_ resampling partitions which are then split into analysis and assessment sets of size $n_{fit}$ and $n_{pred}$, respectively. It turns out that, as $n_{pred}$ becomes larger, the bias increases in a pessimistic direction^[Meaning that performance looks worse than it should. For example, smaller $R^2$ or inflated Brier score.]. In other words, if our training set has 1,000 samples, a resampling method where the assessment set has $n_{pred} = 100$ data points has a smaller bias than one using $n_{pred} = 500$. Another thing that we know is that increasing the number of resamples (_B_) can't significantly reduce the bias.  Therefore, we must think carefully about the way the data is partitioned in the resamplint process to reduce potential bias.
```

```{r}
#| label: fig-mccv-stats
#| echo: false
#| fig-width: 6
#| fig-height: 3.5
#| fig-align: "center"
#| out-width: "70%"
#| fig-cap: Variance and bias statistics for simulated data using different configurations for Monte Carlo cross-validation. The range of the y-axes are common across similar plots below for different resampling techniques.  

mc_bias <-
  stats_mc_cv %>%
  filter(model == sel_model & .metric == sel_metric) %>%
  ggplot(aes(times, bias, col = retain, pch = retain)) +
  geom_point() +
  geom_line() +
  geom_hline(yintercept = 0, lty = 2) +
  scale_y_continuous(labels = scales::percent, limits = rng_bias) +
  scale_color_brewer(palette = "Set1") +
  labs(y = "Bias", x = "Number of Resamples")

mc_prec <-
  stats_mc_cv %>%
  filter(model == sel_model & .metric == sel_metric) %>%
  ggplot(aes(times, std_err, col = retain, pch = retain)) +
  geom_point() +
  geom_line() +
  lims(y = rng_std_err) +
  geom_hline(yintercept = 0, lty = 2) +
  labs(y = "Std. Error", x = "Number of Resamples") +
  scale_color_brewer(palette = "Set1")

mc_bias + mc_prec +
  plot_layout(guides = 'collect') &
  theme(legend.position = "top")
```

> The results of this simulation indicate that, in terms of precision, there is little benefit in including more than 70% of the training set in the assessment set. Also, while more resamples always help, there is incremental benefit in using more than 50 or 60 resamples (for this size training set). Regarding bias, the decrease somewhat slows down near 80% of the training set retained. Holdout proportions of around 75% - 80% might be a reasonable rule of thumb (these are, of course, subjective).

Reviewer: ```suggestion
The results of this simulation indicate that, in terms of precision, there is little benefit in including more than 70% of the training set in the assessment set. Also, while more resamples always help, there is incremental benefit in using more than 50 or 60 resamples (for this size training set). Regarding bias, the decrease somewhat slows down near 80% of the training set retained. Holdout proportions of around 75% - 80% might be a reasonable rule of thumb (these are, of course, subjective and based on this simulation).
```
